[DTensor] Fix default_strategy and rename for clarity

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #158538
* __->__ #158490

Fixes several bugs in the original.
- foremost, fixes a serious bug where we returned incorrect strategies
  by mixing input_specs that were frozen from
  select_strategy.strategies[0] with output_specs that varied across
  select_strategy.strategies[0..N] (e.g. we could create a nonsense
  strategy like input:Shard(0) output(Replicate) for an op like clone
- fixes the redistribute costs: they should not actually be 0, they
  should be the cost of redistributing our single input from another
  strategy to the current strategy, in our list of output strategies
- adds a note, wondering if we should have just literally returned the
  input strategy instead of creating this new object
- Currently, using default_strategy is incorrect becuase it maps 'self'
  tensor's strategies directly onto 'src' tensor without accounting for
  the fact that copy_ supports broadcasting a smaller rank tensor into a
  larger one.

Separates out copy_  op from default strategy, adds missing test case,
but does not fix the underlying issue with copy_, leaves that for future
PR

Renames to `propagate_single_input_strategy` since that's more
descriptive

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @d4l3k @pragupta

wconstab committed:
[DTensor] Fix default_strategy and rename for clarity

Fixes several bugs in the original.
- foremost, fixes a serious bug where we returned incorrect strategies
  by mixing input_specs that were frozen from
  select_strategy.strategies[0] with output_specs that varied across
  select_strategy.strategies[0..N] (e.g. we could create a nonsense
  strategy like input:Shard(0) output(Replicate) for an op like clone
- fixes the redistribute costs: they should not actually be 0, they
  should be the cost of redistributing our single input from another
  strategy to the current strategy, in our list of output strategies
- adds a note, wondering if we should have just literally returned the
  input strategy instead of creating this new object

Renames to `propagate_single_input_strategy` since that's more
descriptive

[ghstack-poisoned]

zpcore commented on torch/distributed/tensor/_ops/_tensor_ops.py:
```diff
@@ -82,12 +85,12 @@ def default_strategy(op_schema: OpSchema) -> StrategyType:
         aten.zero_.default,
     ]
-)(default_strategy)
+)(propagate_single_input_strategy)
```
Doesn't look like all ops here only have one input arg, e.g., `aten.fill_.Scalar`.

wconstab replied:
trying this instead
`assert len([s for s in op_schema.args_schema if isinstance(s, OpStrategy)]) == 1`

zpcore replied:
Yes, one Tensor input should hold. LGTM now!

zpcore replied:
Oh, wait, `func: copy_(Tensor(a!) self, Tensor src, bool non_blocking=False) -> Tensor(a!)` still have two.

wconstab committed:
Update on "[DTensor] Fix default_strategy and rename for clarity"


Fixes several bugs in the original.
- foremost, fixes a serious bug where we returned incorrect strategies
  by mixing input_specs that were frozen from
  select_strategy.strategies[0] with output_specs that varied across
  select_strategy.strategies[0..N] (e.g. we could create a nonsense
  strategy like input:Shard(0) output(Replicate) for an op like clone
- fixes the redistribute costs: they should not actually be 0, they
  should be the cost of redistributing our single input from another
  strategy to the current strategy, in our list of output strategies
- adds a note, wondering if we should have just literally returned the
  input strategy instead of creating this new object

Renames to `propagate_single_input_strategy` since that's more
descriptive

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Fix default_strategy and rename for clarity"


Fixes several bugs in the original.
- foremost, fixes a serious bug where we returned incorrect strategies
  by mixing input_specs that were frozen from
  select_strategy.strategies[0] with output_specs that varied across
  select_strategy.strategies[0..N] (e.g. we could create a nonsense
  strategy like input:Shard(0) output(Replicate) for an op like clone
- fixes the redistribute costs: they should not actually be 0, they
  should be the cost of redistributing our single input from another
  strategy to the current strategy, in our list of output strategies
- adds a note, wondering if we should have just literally returned the
  input strategy instead of creating this new object

Renames to `propagate_single_input_strategy` since that's more
descriptive

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Fix default_strategy and rename for clarity"


Fixes several bugs in the original.
- foremost, fixes a serious bug where we returned incorrect strategies
  by mixing input_specs that were frozen from
  select_strategy.strategies[0] with output_specs that varied across
  select_strategy.strategies[0..N] (e.g. we could create a nonsense
  strategy like input:Shard(0) output(Replicate) for an op like clone
- fixes the redistribute costs: they should not actually be 0, they
  should be the cost of redistributing our single input from another
  strategy to the current strategy, in our list of output strategies
- adds a note, wondering if we should have just literally returned the
  input strategy instead of creating this new object

Renames to `propagate_single_input_strategy` since that's more
descriptive

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

wanchaol commented on torch/distributed/tensor/_ops/_tensor_ops.py:
```diff
@@ -39,55 +39,98 @@
+                input_specs=[
+                    DTensorSpec(
+                        mesh=first_input_strategy.mesh,
+                        placements=strategy.output_spec.placements,
```
The intention of default strategy means to let every argument follow the input argument strategy, but here it looks like you only assume the op have a single input. Probably make sense for the renaming you have, but it was not the same intended behavior as before

wanchaol commented on torch/distributed/tensor/_ops/_tensor_ops.py:
```diff
@@ -39,55 +39,98 @@
-                tensor_meta=strategy.output_spec.tensor_meta,
-            ),
-            input_specs=input_specs,
-            redistribute_cost=redistribute_cost,
```
aren't the real fix here should simply be generating the redistribute_cost here by calling                `generate_redistribute_costs`?

wanchaol reviewed (changes requested):
Could you explain more about the bug? I thought it could be easily fixed by dynamically generate redistribute_costs. This changes the intention of default_strategy and I wonder if it's necessary

wconstab committed:
Update on "[DTensor] Fix default_strategy and rename for clarity"


Fixes several bugs in the original.
- foremost, fixes a serious bug where we returned incorrect strategies
  by mixing input_specs that were frozen from
  select_strategy.strategies[0] with output_specs that varied across
  select_strategy.strategies[0..N] (e.g. we could create a nonsense
  strategy like input:Shard(0) output(Replicate) for an op like clone
- fixes the redistribute costs: they should not actually be 0, they
  should be the cost of redistributing our single input from another
  strategy to the current strategy, in our list of output strategies
- adds a note, wondering if we should have just literally returned the
  input strategy instead of creating this new object

Renames to `propagate_single_input_strategy` since that's more
descriptive

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Fix default_strategy and rename for clarity"


Fixes several bugs in the original.
- foremost, fixes a serious bug where we returned incorrect strategies
  by mixing input_specs that were frozen from
  select_strategy.strategies[0] with output_specs that varied across
  select_strategy.strategies[0..N] (e.g. we could create a nonsense
  strategy like input:Shard(0) output(Replicate) for an op like clone
- fixes the redistribute costs: they should not actually be 0, they
  should be the cost of redistributing our single input from another
  strategy to the current strategy, in our list of output strategies
- adds a note, wondering if we should have just literally returned the
  input strategy instead of creating this new object
- Currently, using default_strategy is incorrect becuase it maps 'self'
  tensor's strategies directly onto 'src' tensor without accounting for
  the fact that copy_ supports broadcasting a smaller rank tensor into a
  larger one.

Separates out copy_  op from default strategy, adds missing test case,
but does not fix the underlying issue with copy_, leaves that for future
PR

Renames to `propagate_single_input_strategy` since that's more
descriptive

[ghstack-poisoned]

XilunWu commented on torch/distributed/tensor/_ops/_tensor_ops.py:
```diff
@@ -39,55 +39,98 @@
+    #
+    # For now, I just keep copy working essentially the way it was before this PR,
+    # but split it out so it can be handled separately in the future.
+    num_tensor_args = 2
+    first_input_strategy = op_schema.args_schema[0]
+    assert isinstance(first_input_strategy, OpStrategy)
+    return OpStrategy(
+        [
+            OpSpec(
+                output_specs=DTensorSpec(
+                    mesh=first_input_strategy.mesh,
+                    placements=strategy.output_spec.placements,
+                    tensor_meta=strategy.output_spec.tensor_meta,
+                ),
+                input_specs=[
+                    DTensorSpec(
+                        mesh=first_input_strategy.mesh,
+                        placements=strategy.output_spec.placements,
+                        tensor_meta=strategy.output_spec.tensor_meta,
+                    )
+                    for _ in range(num_tensor_args)
+                ],
+                redistribute_cost=[
+                    generate_redistribute_costs(
+                        first_input_strategy, strategy.output_spec
+                    )
+                    for _ in range(num_tensor_args)
+                ],
+            )
+            for strategy in first_input_strategy.strategies
+        ]
+    )
```
IIUC this doesn't address the broadcast issue either. Do we address it in another PR?

wconstab replied:
see the next PR in the stack

XilunWu reviewed (commented):
The idea that we need a `single_operand_strategy` looks good to me. Would like discuss more:

The pattern of "all DTensor input should follow a given input's sharding" seems common. I believe we still need it (of course it needs fix) with a new name. But I'm okay with removing it in this PR and add a correct version back in another since the only use case right now is `copy_` and `copy_` needs to handle broadcast. 

BTW, do some `foreach_*` ops need single-input propagate and follow-some-input propagate as well?

XilunWu commented on test/distributed/tensor/test_tensor_ops.py:
```diff
@@ -53,6 +53,38 @@ def test_clone(self):
+            dst_tensor.copy_(src_tensor)
+            self.assertEqual(dst_dtensor.full_tensor(), dst_tensor)
+
+    # @pytest.mark.xfail
+    # @with_comms
+    # def test_copy_broadcast(self):
+    #     device_mesh = DeviceMesh(self.device_type, list(range(self.world_size)))
+    #     src_specs = [[Replicate()], [Shard(0)]]
+    #     src_tensor = torch.randn((12,))
+
+    #     dst_tensor = torch.zeros(12, 12)
+    #     dst_specs = [[Replicate()], [Shard(1)]]
+    #     for dst_spec, src_spec in zip(dst_specs, src_specs):
+    #         src_dtensor = distribute_tensor(src_tensor, device_mesh, dst_spec)
+    #         dst_dtensor = distribute_tensor(dst_tensor, device_mesh, src_spec)
+    #         # perform a broadcasted copy from Shard(0) to Shard(1) for the worst case
+    #         dst_dtensor.copy_(src_dtensor)
+    #         dst_tensor.copy_(src_tensor)
+    #         self.assertEqual(dst_dtensor.full_tensor(), dst_tensor)
+
```
let's remove this

XilunWu reviewed (approved):
The strategy part LGTM. Didn't review the `copy_` part because they're override in next PR.

wanchaol commented on test/distributed/tensor/test_tensor_ops.py:
```diff
@@ -53,6 +53,38 @@ def test_clone(self):
+
+    # @pytest.mark.xfail
+    # @with_comms
+    # def test_copy_broadcast(self):
```
why this being comment out? I thought you are marking xfails

wconstab replied:
xfail did not work haha. so i commented it. its uncommented and passing in the next PR.

wanchaol commented on torch/distributed/tensor/_ops/_tensor_ops.py:
```diff
@@ -39,55 +39,98 @@
-                mesh=select_strategy.mesh,
-                placements=select_strategy.strategies[0].output_spec.placements,
-                tensor_meta=select_strategy.strategies[0].output_spec.tensor_meta,
+def propagate_single_input_strategy(op_schema: OpSchema) -> StrategyType:
```
Oh actually i think it's probably fine, it was added in this PR https://github.com/pytorch/pytorch/commit/ae86e8f6c829a3cfa9204949156fce2d048c919b

Before that default strategy does not handle multiple inputs anyways.

wconstab commented:
@pytorchbot merge

clee2000 commented:
@pytorchbot revert -m "broke lint? [GH job link](https://github.com/pytorch/pytorch/actions/runs/16361950974/job/46231492581) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/d8b084312b54e97bdbaf6a178fe2fc628a23243b)" -c landrace

cc @seemethere tentatively marking this as landrace but the misspelling does exist in the PR as well

wconstab committed:
Update on "[DTensor] Fix default_strategy and rename for clarity"


Fixes several bugs in the original.
- foremost, fixes a serious bug where we returned incorrect strategies
  by mixing input_specs that were frozen from
  select_strategy.strategies[0] with output_specs that varied across
  select_strategy.strategies[0..N] (e.g. we could create a nonsense
  strategy like input:Shard(0) output(Replicate) for an op like clone
- fixes the redistribute costs: they should not actually be 0, they
  should be the cost of redistributing our single input from another
  strategy to the current strategy, in our list of output strategies
- adds a note, wondering if we should have just literally returned the
  input strategy instead of creating this new object
- Currently, using default_strategy is incorrect becuase it maps 'self'
  tensor's strategies directly onto 'src' tensor without accounting for
  the fact that copy_ supports broadcasting a smaller rank tensor into a
  larger one.

Separates out copy_  op from default strategy, adds missing test case,
but does not fix the underlying issue with copy_, leaves that for future
PR

Renames to `propagate_single_input_strategy` since that's more
descriptive

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k pragupta

[ghstack-poisoned]
